### PR TITLE
[CAS-437] Refactor TokenManager and handle custom TokenProvider errors

### PR DESCRIPTION
**Pull request by bychkovdmitry**
### Description

This PR fixes the following issues:
1. `tokenManager.loadAsync` callback was executed on a worker thread
```
tokenManager.loadAsync {
     // This code is now executed on the Main thread
     if (it.isSuccess) {
         clientStateService.onTokenReceived(it.data())
     } else {
         clientStateService.onTokenReceived("")
     }
    connect()
}
```
2. `tokenManager.loadAsync` is now guaranteed to be executed only once which was not the case previously as  `TokenManager` class was not thread-safe
3. Retry `TokenProvider.loadToken()` several time in case of errors
4. Cancel lengthy `loadAsyncAndRetry` job before logout so that loaded token does not arrive when we are in the `Idle` state on the login screen (otherwise will crash)

What's not done:
Unclear what to do after several unsuccessful retries. Maybe we need a retry button in our headers.

### Checklist

- [X] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [X] PR targets the `develop` branch
- [ ] Changelog updated with client-facing changes
- [X] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [X] Reviewers added

### CODE REVIEW

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/token/TokenManagerImpl.kt [hunk 1]
```diff
@@ -2,80 +2,89 @@ package io.getstream.chat.android.client.token
...
+    private suspend fun runAndRetry(): String {
+        var attempt = 0
+        var token: String = EMPTY_TOKEN
+
+        while (attempt < ATTEMPT_COUNT) {
+            token = loadSync()
+            if (token != EMPTY_TOKEN) {
+                break
+            } else {
+                delay(RETRY_DELAY_MILLIS)
```
Comment (JcMinarro) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/token/TokenManagerImpl.kt:78]: Could we use an Exponential backoff?

Comment (bychkovdmitry) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/token/TokenManagerImpl.kt:78]: Yes, will do. And later I will review our retry use cases, maybe we can come up with a neat solution.

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/ChatSocketServiceImpl.kt [hunk 2]
```diff
@@ -170,28 +174,36 @@ internal class ChatSocketServiceImpl private constructor(
...
+                }
+                is ConnectionConf.UserConnectionConf -> {
+                    state = State.Connecting
+                    socketConnectionJob = GlobalScope.launch(DispatcherProvider.IO) {
+                        tokenManager.ensureTokenLoaded()
+                        withContext(DispatcherProvider.Main) {
+                            socket = socketFactory.createNormalSocket(endpoint, apiKey, user)
+                        }
+                    }
+                }
```
Comment (bychkovdmitry) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/ChatSocketServiceImpl.kt:200]: The job needs to cancelled as `tokenManager` can load token for quite a while and the following will have no effect
```
 private fun shutdownSocketConnection() {
        socket?.close(1000, "Connection close by client")	       
        socket = null	       
    }	
```

File: stream-chat-android-client/src/test/java/io/getstream/chat/android/client/clientstate/ClientStateServiceTests.kt [hunk 3]
```diff
@@ -45,47 +45,37 @@ internal class ClientStateServiceTests {
     }
 
     @Test
-    fun `Given Idle state When set user Should move to state user pending without token`() {
+    fun `Given Idle state When set user Should move to state user pending`() {
```
Comment (bychkovdmitry) [inline stream-chat-android-client/src/test/java/io/getstream/chat/android/client/clientstate/ClientStateServiceTests.kt:48]: These tests are not that sophisticated now, can be simplified later

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/token/ConstantTokenProvider.kt [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package io.getstream.chat.android.client.token
+
```
Comment (bychkovdmitry) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/token/ConstantTokenProvider.kt:2]: This name seems more descriptive to me

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/ChatSocketServiceImpl.kt [hunk 5]
```diff
@@ -170,28 +174,36 @@ internal class ChatSocketServiceImpl private constructor(
...
+            }.exhaustive
+        }
     }
```
Comment (bychkovdmitry) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/ChatSocketServiceImpl.kt:204]: Either token is loading, then cancel job. Or socket has already been created, then shutdown it

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/errors/ChatErrorCode.kt [hunk 6]
```diff
@@ -16,9 +14,23 @@ public enum class ChatErrorCode(public val code: Int, public val description: St
     UNABLE_TO_PARSE_SOCKET_EVENT(1008, "Socket event payload either invalid or null"),
     NO_ERROR_BODY(1009, "No error body. See http status code"),
 
-    /**
-     * Backend
-     */
+    // server error codes
+    AUTHENTICATION_ERROR(5, "Unauthenticated, problem with authentication"),
     TOKEN_EXPIRED(40, "Token expired, new one must be requested."),
```
Comment (bychkovdmitry) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/errors/ChatErrorCode.kt:19]: Maybe just use 401 http code? 
https://getstream.io/chat/docs/android/api_errors_response/?language=kotlin&q=token

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/ChatSocketServiceImpl.kt [hunk 7]
```diff
@@ -94,31 +100,29 @@ internal class ChatSocketServiceImpl private constructor(
...
-        ChatErrorCode.CANT_PARSE_CONNECTION_EVENT.code,
-        ChatErrorCode.CANT_PARSE_EVENT.code,
-        ChatErrorCode.UNABLE_TO_PARSE_SOCKET_EVENT.code,
-        ChatErrorCode.NO_ERROR_BODY.code -> {
-            // Nothing to do on that case
-        }
-        ChatErrorCode.TOKEN_EXPIRED.code -> {
+    private fun onChatNetworkError(error: ChatNetworkError) {
+        if (ChatErrorCode.isAuthenticationError(error.streamCode)) {
             tokenManager.expireToken()
```
Comment (JcMinarro) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/ChatSocketServiceImpl.kt:105]: Don't we load a new one?

**Review by JcMinarro — APPROVED** [review]
LGTM

**Pull request merged**